### DISCUSSION TITLE
Chart to WPF ImageSource

### DISCUSSION THREAD

**Discussion by jobrow-msft** [Live-Charts/LiveCharts2#1069] [Q&A]
Has anyone had any success converting a chart to an ImageSource that can be used with a drawing context? I have a control that is all programmatically rendered and I would like to draw a chart, but so far I haven't had any luck. Here are some things I've tried:

SKPieChart chart = new SKPieChart(new LiveChartsCore.SkiaSharpView.WPF.PieChart { Series = chartSeries.ToArray() }) { Width = chartSize, Height = chartSize };

SKImage chartImage = chart.GetImage();
SKData encodedImage = chartImage.Encode();

using (Stream stream = encodedImage.AsStream())
{
    var bitmap = new System.Windows.Media.Imaging.BitmapImage();
    bitmap.BeginInit();
    bitmap.StreamSource = stream;
    bitmap.CacheOption = System.Windows.Media.Imaging.BitmapCacheOption.OnLoad;
    bitmap.EndInit();
    bitmap.Freeze();
    dc.DrawImage(bitmap, new Rect(this.LeftEdge, lineRect.Top, chartSize, chartSize));
}

and

SKPieChart chart = new SKPieChart(new LiveChartsCore.SkiaSharpView.WPF.PieChart { Series = chartSeries.ToArray() }) { Width = chartSize, Height = chartSize };

using var memoryStream = new MemoryStream();
var canvas = SKSvgCanvas.Create(SKRect.Create(chart.Width, chart.Height), memoryStream);
chart.DrawOnCanvas(canvas);
canvas.Dispose(); // <- dispose it before using the stream, otherwise the svg could not be completed.

var imageSource = new System.Windows.Media.Imaging.BitmapImage();
imageSource.BeginInit();
imageSource.StreamSource = memoryStream;
imageSource.EndInit(); // This actually causes an exception

dc.DrawImage(imageSource, new Rect(this.LeftEdge, lineRect.Top, chartSize, chartSize));

**Comment by beto-rodriguez** [comment]
I saw #795, where the Autor of the issue used SkiaSharp to draw on top of the image generated by LiveCharts, that is an interesting sample that might help you (and could be also useful to include that in the repo).  [sample here](https://github.com/beto-rodriguez/LiveCharts2/files/10133139/sample-code.txt)

**Reply by jobrow-msft** [reply]
Thanks for the tip! I kept digging into this and found that the code to generate the image source actually works fine, but that both it and SaveImage() give a blank result when I try to use a size of 15x15. If I use 200x200 it all works as expected, but I am trying to render tiny thumbnails that then so a larger chart on hover. Is there a limitation to how small a chart can be drawn?

`SKPieChart chart = new SKPieChart(new LiveChartsCore.SkiaSharpView.WPF.PieChart { Series = chartSeries.ToArray() }) { Width = 15, Height = 15 };
chart.SaveImage(@"<path>");`

**Comment by jobrow-msft** [comment]
For another data point, I found that if I create the chart as 200x200, but draw it as 15x15, I can get it to show up, although it is very pixelated compared to how clean dc.DrawEllipse can draw a circle.

SKPieChart chart = new SKPieChart(new LiveChartsCore.SkiaSharpView.WPF.PieChart { Series = chartSeries.ToArray() }) { Width = 200, Height = 200 };
SKImage chartImage = chart.GetImage();
SKData encodedImage = chartImage.Encode();

using (Stream stream = encodedImage.AsStream())
{
	var bitmap = new System.Windows.Media.Imaging.BitmapImage();
	bitmap.BeginInit();
	bitmap.StreamSource = stream;
	bitmap.CacheOption = System.Windows.Media.Imaging.BitmapCacheOption.OnLoad;
	bitmap.EndInit();
	bitmap.Freeze();
	dc.DrawImage(bitmap, new Rect(x, y, 15, 15));
}

![image](https://github.com/beto-rodriguez/LiveCharts2/assets/115046097/60730bca-2e0a-4d53-bc70-196e9ca532b9)

Still working on finding a way to draw a non-pixelated thumbnail, so I'll leave my findings and progress here as I go in case it helps someone in the future.
